#184752098  User with seller as a Role can Add product

bergad37 opened the pull request:
## What does this PR do?

Creating and adding a product in the stock 

## Description of Task to be completed?

- A seller should be able to add/create only a new product for sale, such that the stock can be seen by any requesting client
- An existing product can not be added in the collection,
- The existing product is fetched based on the name and it's corresponding owner's collection,
- A seller is given the option to edit product details on trying to create existing product in his collection
- Create-product-route can only be accessed by a user signed in as a seller. 


## How should this be manually tested?

- Clone the repo
- Checkout to the branch by `git checkout ft-create-product-184752098` command
- Install dependencies by `npm i`
- Refer to `.env.example` and create `.env` in the root directory.
- Start the application development server: `npm run dev`
- 
- Learn more about the endpoint using the testing tool like [postman](https://www.postman.com/) on this address: http://localhost:{PORT}/api/product/addProduct.
- (http://localhost:{PORT}/api/product/allProduct.)


## Any background context you want to provide?
You can read more about validation and associations between tables [here](https://sequelize.org/docs/v6/core-concepts/assocs/)
## What are the relevant pivotal tracker/Trello stories?
[#184752098](https://www.pivotaltracker.com/story/show/184752098)

niyontwali reviewed (changes requested):
Please rebase your branch to develop

nishimweprince reviewed (changes requested):
Please rebase with develop and update this PR to remove conflicts. @jcharles921 has assigned the label for you.

bergad37 committed:
feat(create) a seller create a product

-seller can add a product in collection
-seller can see all product in the collection

[Delivers #184752098]

niyontwali reviewed (changes requested):
@bergad37 

You should have a middleware of checkActiveUser or whatever name you give it. I am seeing that all the checks where done in the controller. Such check should be in middlewares, and make sure you capture all errors possible i.e on active and on inactive, otherwise if the user is inactive and you did not capture that, your running application will be stopped in the production

nishimweprince commented on src/controller/productController.js:
```diff
@@ -0,0 +1,122 @@
+
+      const decodedToken = jwt.verify(token, process.env.USER_SECRET);
+      const { id } = decodedToken;
+      const { email } = await user.findOne({ where: { email: id } });
```
Given the team is proposing to use IDs instead of emails to sign **JWTs**, you should consider searching the user using ID instead of email on this line.

You will have something like this: `const { email } = await user.findOne({ where: { id } });`

Same applies to all places where you were searching for a user using email attribute.

nishimweprince commented on src/docs/components.js:
```diff
@@ -44,114 +50,99 @@ const components = {
     },
 
     /**
```
Looks like it would comment out all lines after it.

nishimweprince commented on src/test/product.test.js:
```diff
@@ -0,0 +1,127 @@
+
+
+
+// login test
```
Please remove the commented lines.

nishimweprince reviewed (changes requested):
I left some comments. Also rebase to have the latest changes from develop.

nishimweprince reviewed (changes requested):
I suggested some changes.

bergad37 commented:
> @bergad37
> 
> You should have a middleware of checkActiveUser or whatever name you give it. I am seeing that all the checks where done in the controller. Such check should be in middlewares, and make sure you capture all errors possible i.e on active and on inactive, otherwise if the user is inactive and you did not capture that, your running application will be stopped in the production

Okay ,noted! i have now separated the middleware with the controllers.And rebased on `develop` branch, now you can check on that!

niyontwali commented:
Please rebase

bergad37 reviewed (commented):
File changes viewed.
